test/library/packages/VectorList/VectorLists.chpl

```diff
@@ -0,0 +1,217 @@
+      this.eltType = eltType;
+      this.head = new unmanaged VectorListElement(eltType, initCapacity);
+      this.tail = this.head;
+    }
+
+    proc deinit() {
+      // We do not want "owned" for the VectorListElement links, i.e. this.head,
+      // this.head.next, .next.next, etc. Because with "owned" these links would
+      // be deinit-ed with recursive calls to VectorListElement.deinit(), one
+      // call per list element. A long list would lead to a lot of stack frames.
+      // Also, "owned" would not allow us to update this.tail in push_back().
+      for LE in listElements() do
+        delete LE;
+    }      
+  }
+
+  /*
+  Store 'size' elements in 'elements', with 'capacity'.
+  Currently the backing store is allocated right away and does not change.
+  */
+  pragma "no doc"
+  class VectorListElement {
+    type eltType;
+    var size: int;
+    const capacity: int;
+    const elements: _ddata(eltType);  // _ddata is not managed
+    var next: unmanaged VectorListElement(eltType);
+
+    inline proc init(type eltType, initCapacity: int) {
+      this.eltType = eltType;
+      this.size = 0;
+      this.capacity = initCapacity;
+      this.elements = _ddata_allocate(eltType, this.capacity);
+      this.next = nil;
+    }
+
+    proc deinit() {
+      // Deinit individual elements.
+      for i in 0..#size do
+        chpl__autoDestroy(elements[i]);
+
+      _ddata_free(elements, capacity);
+    }
+
+    // 0-based indexing
+    inline proc elemAt(idx: int) ref {
+      return elements.this(idx);
+    }
+  }
+
+  /* Add 'val' to the end of the collection */
+  proc VectorList.push_back(in val: this.eltType) lifetime this < val {
+    const last = this.tail;
+    if boundsChecking then assert(last.next == nil);
+    this.size += 1;
+
+    if (last.size < last.capacity) {
+      // Add to the currently-last VLE.
+      last.elemAt(last.size) = val;
+      last.size += 1;
+      return;
+    }
+
+    // Otherwise append a new VLE.
+    const nextCapacity =
+      if last.capacity * vectorListGrowFactor > vectorListMaxCapacity
+      then last.capacity else last.capacity * vectorListGrowFactor;
+
+    var next = new unmanaged VectorListElement(eltType, nextCapacity);
+    next.elemAt(0) = val;
+    next.size = 1;
+
+    // Update link pointers.
+    last.next = next;
+    this.tail = next;
+  }
+
+  /* Yields all VectorListElement. */
+  inline iter VectorList.listElements() {
+    var curr = head;
+    do { const next = curr.next; yield curr; curr = next; }
+    while curr != nil;
+  }
+
+  /* Yields pairs (starting index, VectorListElement pointer). */
+  inline iter VectorList.links(): (int,VectorListElement(eltType)) {
+    if size == 0 then return; // no vector elements
+    var currStart = 0;
+    for LE in listElements() {
+      yield (currStart, LE);
+      currStart += LE.size;
+    }
+    if boundsChecking then assert(currStart == size);
+  }
+
+  /* Indexing is O(log(size)) time. */
```

Comment:
Indexing can be O(1) with this sort of strategy. See 

https://github.com/mppf/chapel/commit/14632d69688becfc5ae6f446095a461f6c5e1528#diff-4514f4137c9f21d19556b229b6fd35fdR186
https://github.com/mppf/chapel/commit/14632d69688becfc5ae6f446095a461f6c5e1528#diff-05924e5417bb0cfed1243f49cf29563dR380

or the paper "Singly Resizable Arrays and Deques" section of Brodnik Carlsson Demaine Munro Sedgewick  "Resizable Arrays in Optimal Time and Space" which gives O(1)  indexing and O(sqrt(n)) space overhead with O(sqrt(n)) blocks.

Reply:
I like this idea. Or we can set up O(1)-time indexing with log(n) or O(number of blocks) memory overhead.